File: middle/5-longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
  if (s === null || s.length === 0) return '';
  let start = 0, end = 0;
  for(let i = 0; i < s.length; i++) {
    let len1 = expandCenter(s, i, i);
    let len2 = expandCenter(s, i, i+1);
    let len = Math.max(len1, len2);
    if (len > end - start) {
      start = i - (len - 1) / 2;
      end = i + len / 2;
    }
  }
  return s.substring(start, end + 1);
};

function expandCenter(s, left, right) {
  let L = left;
  let R = right;
  while (L >= 0 && R < s.length && s[L] === s[R]){
    L--;
    R++;
  }
  return R - L - 1;
}

console.log(longestPalindrome('dababad'));

File: middle/5-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        if s == '':
          return ''
        start = 0
        end = 0
        for i in range(len(s)):
          len1 = self.expandCenter(s, i, i)
          len2 = self.expandCenter(s, i, i+1)
          lenMax = max(len1, len2)
          if lenMax > end - start:
            start = i - (lenMax - 1) / 2
            end = i + lenMax / 2
        return s[start:end + 1]

    def expandCenter(self, s, start, end):
      L = start
      R = end
      while L >= 0 and R < len(s) and s[L] == s[R]:
        L -= 1
        R += 1
      return R - L - 1

